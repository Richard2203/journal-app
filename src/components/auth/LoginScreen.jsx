import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { useDispatch } from 'react-redux';
import { login, startLoginEmailPassword } from '../../action/auth';

const LoginScreen = () => {
	const [formValues, handleInputChange] = useForm({
		email: 'alpaca@outlook.com',
		password: 'alpaca123',
	});

	// el custom hook useDispatch proviene de redux y
	// sirve para retornar una funcion que recibe un metodo mediante el cual
	// tendremos acceso a Dispatch enviandole informacion
	const dispatch = useDispatch();

	const { email, password } = formValues;

	const handleLogin = (e) => {
		e.preventDefault();
		dispatch(startLoginEmailPassword(email, password));
	};

	return (
		<>
			<h3 className="auth__title">Login</h3>
			<form onSubmit={handleLogin}>
				<div className="form-control">
					<input
						type="text"
						id="email"
						name="email"
						className="form-control__input"
						autoComplete="off"
						placeholder=" "
						value={email}
						onChange={handleInputChange}
					/>

					<label htmlFor="email" className="form-control__label">
						Email
					</label>
				</div>

				<div className="form-control">
					<input
						type="password"
						id="password"
						name="password"
						className="form-control__input"
						autoComplete="off"
						placeholder=" "
						value={password}
						onChange={handleInputChange}
					/>

					<label htmlFor="password" className="form-control__label">
						password
					</label>
				</div>

				<button type="submit" className="btn cursor btn-block">
					Login
				</button>
			</form>

			<hr />

			<h6 className="auth__title-network">Login with social Networks</h6>
			<div className="google-btn">
				<div className="google-icon-wrapper">
					<img
						className="google-icon"
						src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
						alt="google button"
					/>
				</div>
				<p className="btn-text">
					<b>Sign in with google</b>
				</p>
			</div>

			<Link to="/auth/register" className="link">
				Create a new account
			</Link>
		</>
	);
};

export default LoginScreen;
