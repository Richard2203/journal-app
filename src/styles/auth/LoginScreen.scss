*,
*::before,
*::after {
	box-sizing: border-box;
	margin: 0;
	padding: 0;
}

body {
	font-family: 'Convergence', sans-serif;
	display: grid;
	justify-content: center;
	align-items: center;
	font-size: 1.2rem;
}

.form-control {
	position: relative;
	width: 20rem;
	height: 3rem;

	&__input {
		position: absolute;
		top: 0;
		left: 0;
		width: 100%;
		height: 100%;
		border: 2px solid $clr-gray300;
		border-radius: $radius;
		font-family: inherit;
		font-size: inherit;
		color: black;
		outline: none;
		padding: 1.25rem;
		background: none;

		&:hover {
			border-color: $clr-primary-light;
		}

		&:focus {
			border-color: $clr-primary;
		}
	}

	&__label {
		position: absolute;
		left: 1rem;
		top: 0.8rem;
		padding: 0 0.5rem;
		color: black;
		cursor: text;
		transition: top 200ms ease-in, left 200ms ease-in,
			font-size 200ms ease-in;
		background-color: $light-gray;
	}
}

.form-control__input:focus ~ .form-control__label,
.form-control__input:not(:placeholder-shown).form-control__input:not(:focus)
	~ .form-control__label {
	top: -0.5rem;
	font-size: 0.8rem;
	left: 0.8rem;
}

.btn {
	padding: 7px 14px;
	border-radius: $radius;
	outline: none;
	border: 2px solid $light-blue;

	&:hover {
		border: 3px solid $clr-primary-light;
	}

	&:focus {
		border: 3px solid $clr-primary;
	}
}
