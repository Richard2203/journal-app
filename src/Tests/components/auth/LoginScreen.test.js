// Instalando la dependencia para poder hacer pruebas mock en store de redux
//  https://www.npmjs.com/package/redux-mock-store
// se aÃ±aden las importaciones
//  import configureStore from 'redux-mock-store';
//  const middlewares = [thunk];
//  const mockStore = configureStore(middlewares);
//  const store = mockStore({});

import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
//importamos el middleware que usaremos para pruebas
import thunk from 'redux-thunk';
import {
	startGoogleLogin,
	startLoginEmailPassword,
} from '../../../action/auth';

import LoginScreen from '../../../components/auth/LoginScreen';

// retornamos la propiedad startGoogleLogin que fue asignada como una funcion
// de jest
jest.mock('../../../action/auth', () => ({
	startGoogleLogin: jest.fn(),
	startLoginEmailPassword: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
	auth: {},
	ui: { msgError: null, loading: false },
};

// mocksStore recibe objetos para establecer el store; en este caso
// el store contiene el objeto: auth
let store = mockStore(initState);

// le asignamos un jest.fn() a la funcion dispatch para pdoer controlar
// su comportamiento
store.dispatch = jest.fn();

describe('Pruebas en el componente LoginScreen', () => {
	// Para poder realizar pruebas en redux store se emplean mocks
	//      npm install redux-mock-store --save-dev

	// store ira almacenando todos los getActions() disparados, para evitar esto
	// se emplea beforeEch()
	beforeEach(() => {
		store = mockStore(initState);
		jest.clearAllMocks();
	});

	// - Debido a que LoginScreen emplea Redux debemos pasarle la creacion
	//   de los stores mediante Provider y su propieadad store
	// - Debido a que se emplea la etiqueta Link se deben pasar las referencias
	//   a este componente mediante MemoryRouter, el cual hara de simulador de esta
	//   funcion
	const wrapper = mount(
		<Provider store={store}>
			<MemoryRouter>
				<LoginScreen />
			</MemoryRouter>
		</Provider>
	);

	test('Debe renderizar correctamente el componente', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('debe de ejecutar la funcion startGoogleLogin() en el dispatch', () => {
		wrapper.find('.google-btn').prop('onClick')();
		expect(startGoogleLogin).toHaveBeenCalled();
	});

	test('debe de ejecutar la funcion startLoginEmailPassword() en el dispatch', () => {
		wrapper.find('form').prop('onSubmit')({ preventDefault: () => {} });

		expect(startLoginEmailPassword).toHaveBeenCalled();
		expect(startLoginEmailPassword).toHaveBeenCalledWith('', '');
	});
});
