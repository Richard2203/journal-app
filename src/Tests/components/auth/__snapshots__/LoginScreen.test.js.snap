// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente LoginScreen Debe renderizar correctamente el componente 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Login
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-control"
    >
      <input
        autoComplete="off"
        className="form-control__input"
        id="email"
        name="email"
        onChange={[Function]}
        placeholder=" "
        type="text"
        value=""
      />
      <label
        className="form-control__label"
        htmlFor="email"
      >
        Email
      </label>
    </div>
    <div
      className="form-control"
    >
      <input
        autoComplete="off"
        className="form-control__input"
        id="password"
        name="password"
        onChange={[Function]}
        placeholder=" "
        type="password"
        value=""
      />
      <label
        className="form-control__label"
        htmlFor="password"
      >
        password
      </label>
    </div>
    <button
      className="btn cursor btn-block"
      disabled={false}
      type="submit"
    >
      Login
    </button>
  </form>,
  <hr />,
  <h6
    className="auth__title-network"
  >
    Login with social Networks
  </h6>,
  <div
    className="google-btn"
    onClick={[Function]}
  >
    <div
      className="google-icon-wrapper"
    >
      <img
        alt="google button"
        className="google-icon"
        src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
      />
    </div>
    <p
      className="btn-text"
    >
      <b>
        Sign in with google
      </b>
    </p>
  </div>,
  <a
    className="link"
    href="/auth/register"
    onClick={[Function]}
  >
    Create a new account
  </a>,
]
`;
